### PR TITLE
Napatech bypass v1.03

### PR DESCRIPTION
**Pull request by pjyoung1**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [x] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [x] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to redmine ticket:
https://redmine.openinfosecfoundation.org/issues/2818

Describe changes:
- Added hardware based bypass support on Napatech SmartNICs
- Added support for inline processing using Napatech SmartNICs

This PR Addresses issues identified in review of [PR 4343](https://github.com/OISF/suricata/pull/4343)
  

[PRScript](https://redmine.openinfosecfoundation.org/projects/suricata/wiki/PRscript) output (if applicable):

### CODE REVIEW

File: src/runmode-napatech.c [hunk 1]
```diff
@@ -103,8 +111,24 @@ static int NapatechRegisterDeviceStreams(void)
...
+    if (ConfGetBool("napatech.hardware-bypass", &use_hw_bypass) == 0) {
+        SCLogInfo("napatech.hardware-bypass not found in config file.  Defaulting to disabled.");
+    }
+
+    /* use_all_streams uses existing streams created prior to starting Suricata.  auto_config
+     * automatically creates streams.  Therefore, these two options are mutually exclusive.
+     */
     if (use_all_streams && auto_config) {
-        SCLogError(SC_ERR_RUNMODE, "auto-config cannot be used with use-all-streams.");
+        SCLogError(SC_ERR_RUNMODE, "napatech.auto-config cannot be used in configuration file at the same time as napatech.with use-all-streams.");
```
Comment (jlucovsky) [inline src/runmode-napatech.c]: s/napatech.with use-all-streams/napatech.use-all-streams

File: src/runmode-napatech.c [hunk 2]
```diff
@@ -103,8 +111,24 @@ static int NapatechRegisterDeviceStreams(void)
...
+    }
+
+    /* to use hardware_bypass we need to configure the streams to be consistent.
+     * with the rest of the configuration.  Therefore auto_config is not a valid
+     * option.
+     */
+    if (use_hw_bypass && auto_config == 0) {
+        SCLogError(SC_ERR_RUNMODE, "napatech auto-config must be enabled when using napatech.use-all-streams.");
```
Comment (jlucovsky) [inline src/runmode-napatech.c]: s/use-all-streams/hardware-bypass/ ?

File: src/source-napatech.c [hunk 3]
```diff
@@ -157,6 +172,315 @@ void TmModuleNapatechDecodeRegister(void)
...
+        if (is_inline) {
+            flow_match.keySetId = NAPATECH_FLOWTYPE_PASS;
+        } else {
+            flow_match.keySetId = NAPATECH_FLOWTYPE_DROP;
+        }
+    }
+
+    status = NT_FlowWrite(*phFlowStream, &flow_match, 200);
+    if (status == NT_STATUS_TIMEOUT) {
+        SCLogInfo("NT_FlowWrite returned NT_STATUS_TIMEOUT");
```
Comment (jlucovsky) [inline src/source-napatech.c]: Is a timeout considered success or failure?

File: src/source-napatech.c [hunk 4]
```diff
@@ -369,16 +741,22 @@ TmEcode NapatechPacketLoopZC(ThreadVars *tv, void *data, void *slot)
...
+                status == NT_STATUS_TIMEOUT || status == NT_STATUS_TRYAGAIN)) {
             continue;
         } else if (unlikely(status != NT_SUCCESS)) {
             NAPATECH_ERROR(SC_ERR_NAPATECH_OPEN_FAILED, status);
-            SCLogInfo("Failed to read from Napatech Stream%d: %s",
+            SCLogInfo("Failed to read from Napatech Stream %d: %s",
                     ntv->stream_id, error_buffer);
-            SCReturnInt(TM_ECODE_FAILED);
+            break;
+            //SCReturnInt(TM_ECODE_FAILED);
```
Comment (jlucovsky) [inline src/source-napatech.c]: Delete unused line.

File: src/util-napatech.h [hunk 5]
```diff
@@ -14,43 +14,51 @@
...
-
 #ifndef __UTIL_NAPATECH_H__
 #define __UTIL_NAPATECH_H__
 
 #ifdef HAVE_NAPATECH
 #include <nt.h>
 
-typedef struct NapatechPacketVars_ {
+typedef struct NapatechPacketVars_
+{
```
Comment (jlucovsky) [inline src/util-napatech.h]: nit: suggest consistent placement of `{` on same line with `typedef`

File: src/source-napatech.c (hunk 4 above)
Comment (pjyoung1) [reply src/source-napatech.c]: Fixed

File: doc/userguide/capture-hardware/napatech.rst [hunk 6]
```diff
@@ -232,27 +232,151 @@ Next execute those command using the ntpl tool::
...
+ 
+- napa_total.overflow_drop_pkts - The number of packets that were dropped because 
+  the host buffers were full.  (I.e. the application is not able to process
+  packets quickly enough.)
+ 
+- napa_total.overflow_drop_byte - The number of bytes that were dropped because 
+  the host buffers were full.  (I.e. the application is not able to process
+  packets quickly enough.)
+ 
+On flow aware products the following counters are also available:
```
[the remark this answers is not in the document]
Comment (pjyoung1) [reply doc/userguide/capture-hardware/napatech.rst]: Fixed.

File: src/runmode-napatech.c (hunk 1 above)
Comment (pjyoung1) [reply src/runmode-napatech.c]: Fixed

File: src/runmode-napatech.c (hunk 2 above)
Comment (pjyoung1) [reply src/runmode-napatech.c]: Fixed.

File: src/source-napatech.c (hunk 3 above)
Comment (pjyoung1) [reply src/source-napatech.c]: Added comment and a check for a more serious error.

File: src/util-napatech.h (hunk 5 above)
Comment (pjyoung1) [reply src/util-napatech.h]: Fixed.

**Pull request closed without merging**